flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -1514,8 +1517,10 @@ public void run() {
 					}
 				}
 
-				// send the initial interruption signal
-				executer.interrupt();
+				// send the initial interruption signal, if requested
+				if (invokable.shouldInterruptOnCancel()) {
+					executer.interrupt();
```

Comment:
This interrupt could also slip through in the case as  StefanRRichter mentioned above. Why not use a locked code block to guarantee it?